충남대 BE_이은경 4주차 과제 (3단계)

pkyung opened the pull request:
## 과제 진행 사항
### 옵션 수량 차감
- [x] 수량 차감 메서드 생성
- [x] 얼만큼 양을 뺄 건지의 dto
- [x] 부족한 수량에 대한 exception 생성

pkyung committed:
Chore: 코드 옮기기 - step0

pkyung committed:
Docs: 기능 목록 작성 - step1

pkyung committed:
Feat: 카테고리 도메인 설계

카테고리 엔티티 설계를 구현했다. category의 필드는 id, name, color,
imageUrl, description으로 이뤄져 있으며 모든 필드는 null이 불가하도록
설정했다.

pkyung committed:
Feat: 카테고리의 관계 설정

카테고리는 상품과 1대1 관계라고 생각했다. 따라서, Product 도메인에
@OneToOne 관계로 설정을 해주었다. Category에는 아직 상품을 확인할 필요가
없을 것 같아서 양방향 매핑을 하지 않았다.

pkyung committed:
Feat: 카테고리 ru 구현

카테고리는 상품과 함께 지정할 수 있다는 요구사항이 있다. 따라서 Product
생성자와 update 를 category를 넣은 채로 수정했다.
그리고 category는 수정이 가능하고 모든 category를 조회하는 api가
요구사항에 존재한다. 따라서, findAllCategories와 updateCategory 메서드를
만들어서 카테고리를 조회하고 수정하도록 했다.

pkyung committed:
Refactor: handler 클래스 정리

notfound, alreadyExists, MethodArgument, Unauthorized exception 들을
다루는 handler로 클래스를 정리했다.

pkyung committed:
Chore: 안 옮긴 test 코드 옮김

pkyung committed:
Fix: 상품에 카테고리는 필수로 들어가야 해서 수정함

pkyung committed:
Fix: @ManyToOne으로 수정함

pkyung committed:
Feat: 화면 반영 - 상품과 함께 카테고리  추가되도록

Product를 Page<Product>로 반환하면 Product만 반환되고 연관관계 설정된
Category는 나오지 않아서 ProductResponse를 생성하여 반환하도록 수정헸다.
Product와 Category에 필요한 getter들을 생성했다.

pkyung committed:
Refactor: 필요없는 import 삭제

pkyung committed:
Refactor: 하나만 조회하는 api 또한 productResponse 반환

pkyung committed:
Refactor: token 관련 메서드 import 해서 사용

pkyung committed:
Fix: wish.js의 createWish 전역 변수 설정

pkyung committed:
Refactor: category getter 추가

pkyung committed:
Test: 상품과 카테고리 저장 테스트

pkyung committed:
Docs: 기능 목록 구현 - step1

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Refactor: merge 시 삭제 안된 파일 삭제

pkyung committed:
Refactor: 불필요한 import 제거

pkyung committed:
Refactor: merge 시 삭제 안된 파일 삭제

pkyung committed:
Refactor: category controller와 service 리펙토링

1. requestMapping에 /api/categories 로 수정했다.
2. 메서드명 updateCategory -> update
3. category list의 변수병을 복수형으로 수정했다.
4. update 의 경우 반환할 때 no_content
5. body가 없는 responseEntity는 Void 제네릭 타입으로 명시했다.

pkyung committed:
Refactor: request validation 추가

1. ProductRequest와 CategoryRequest의 color 필드를 정규식으로 검증했다.
2. imageUrl 필드를 정규식을 통해 검증했다.

pkyung committed:
Refactor: productResponse 계층형으로 수정

1. categoryName, categoryId를 category 객체로 수정했다.
2. service의 반환값의 생성자 일부를 수정했다.

pkyung committed:
Refactor: category name 유니크 추가

pkyung committed:
Refactor: orElseThrow로 optional 뗌

pkyung committed:
Refactor: category 엔티티 validate 설정 추가

pkyung committed:
Refactor: description 필드 255자 이상을 위해 @Lob 추가

pkyung committed:
Refactor: Exception 관리

1. AlreadyExistsException과 NotFoundException 을 생성했다.
2. Service에 사용된 Exception들을 수정했다.
3. GlobalExceptionHandler에서 exception을 관리하도록 수정했다.

pkyung committed:
Refactor: 메서드 이름 수정

pkyung committed:
Refactor: Optional orElseThrow로 뗌

pkyung committed:
Docs: 기능 목록 작성 - step2

pkyung committed:
Feat: 옵션 도메인 설계

option의 validation과 필드를 설계했다.

pkyung committed:
Refactor: length가 있기 때문에 notnull이 의미없음

pkyung committed:
Feat: 옵션 dto validation 설정

옵션의 name 필드는 특정한 특수문자만 들어갈 수 있어서 정규식을 통해
설정했다. 그리고 quatity 필드는 최소 1 최대 1억 미만이므로 min max
설정을 했다.

pkyung committed:
Feat: 상품과 옵션 관계 설정

옵션과 상품은 ManyToOne의 관계이다. 따라서, 옵션에는 ManyToOne 매핑을,
상품에는 OneToMany 옵션을 추가했다. 그리고 옵션에 product와의 관계
설정을 위한 setProduct 메서드를 추가하여 product의 options 리스트에
객체를 추가하며 관계를 설정하도록 구현했다.

pkyung committed:
Fix: responseBody와 valid 어노테이션 추가

pkyung committed:
Feat: 옵션 이름 중복 확인 메서드 생성

1. 옵션에 getName 메서드 추가했다.
2. product에 options 필드의 name과 동일한 이름이 있는지 확인하는
   checkDuplicateOptionName 메서드를 추가했다.

pkyung committed:
Feat: 옵션 cr 구현

옵션을 생성하고 조회하도록 구현했다.
생성의 경우, productId와 함께 입력 받아 옵션 생성과 관계 설정을 함께
하도록 구현했다.

pkyung committed:
Fix: pathVariable 인식 안되어 문자열 추가

pkyung committed:
Refactor: imageUrl 정규식 수정함

pkyung committed:
Fix: pathvariable 매개변수 추가

pkyung committed:
Fix: 데이터 직렬화를 위해 getter 추가

pkyung committed:
Refactor: 필요 없는 import 제거

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Docs: 기능 목록 구현 - step2

pkyung committed:
Test: 옵션 테스트 추가

pkyung committed:
Refactor: 멘토님 피드백 추가

checkDuplicateOptionName 메서드를 option에 추가했다. 그리고 stream을
통해 더 간결하게 작성했습니다.
option과 product 사이의 관계를 생성자를 통해 설정했다.

pkyung committed:
Fix: checkDuplicate 위치 재변경

pkyung committed:
Fix: 재변경

pkyung committed:
Test: option service 테스트 구현

option service의 getOptionByProductId와 makeOption을 테스트했다.

pkyung committed:
Refactor: stream으로 수정함

pkyung committed:
Refactor: 테스트 패키지 구조 변경

pkyung committed:
Docs: 기능 목록 생성 - step3

pkyung committed:
Feat: 수량 차감 메서드 생성

subtract 이라는 메서드를 추가하여 매개변수로 받은 값만큼 quantity에서
뺏을 때 0보다 크거나 같으면 처리되도록 구현했다.

pkyung committed:
Feat: 얼만큼의 양을 뺼 건지의 dto

dto에는 id 필드와 quantity 필드가 있다.

pkyung committed:
Feat: 부족한 수량에 대한 exception 생성

부족한 수량을 계산하는 subtract 메서드에서 계산한 값이 0보다 작을 경우에
반환할 에러에 대해서 custom exception을 생성했다. exception이 던져지면
409 conflict 에러를 반환한다.

pkyung committed:
Docs: 기능 목록 구현 - step3

pkyung committed:
Refactor: 단어 일부 수정

pkyung committed:
Merge branch 'pkyung' into main

pkyung committed:
Refactor: 머지 시 삭제 안된 파일 삭제

pkyung committed:
Test: subtract 테스트

pkyung committed:
Fix: 무조건 에러를 던져서 수정함

ksy90101 commented on src/main/java/gift/dto/OptionSubtractRequest.java:
```diff
@@ -0,0 +1,10 @@
+
+import jakarta.validation.constraints.Min;
+
+public record OptionSubtractRequest(
```
Null 체크는 괜찮을까요?

pkyung replied:
저번에 json 형식으로 입력 받을 때 정수형은 null이 불가능하다고 하셔서 하지 않았습니다!

ksy90101 replied:
primery Type vs Boxed Type의 차이를 한번 알아보시면 좋을거 같습니다.

참고링크 : https://stackoverflow.com/questions/2382058/unboxing-null-object-to-primitive-type-results-in-nullpointerexception-fine

ksy90101 commented on src/main/java/gift/exception/OutOfStockException.java:
```diff
@@ -0,0 +1,7 @@
+    public OutOfStockException(String message) {
+        super(message);
+    }
+}
```
```suggestion
}

```

ksy90101 commented on src/main/java/gift/model/Option.java:
```diff
@@ -53,4 +54,12 @@ public Long getQuantity() {
     }
+
+    public void subtract(Long quantity) {
+        if (this.quantity - quantity >= 0) {
```
this.quantity - quantity
이 부분을 변수로 지정해두면 어떨까요?
그럼 훨씬 더 깔끔한 코드가 될거 같습니다.

pkyung replied:
네 변수로 지정하도록 하겠습니다.

ksy90101 commented on src/main/java/gift/model/Option.java:
```diff
@@ -53,4 +54,12 @@ public Long getQuantity() {
     }
+
+    public void subtract(Long quantity) {
+        if (this.quantity - quantity >= 0) {
```
if(new quantity < 0) {
            throw new OutOfStockException("해당 옵션의 수량이 부족합니다.");
}
이런식의 코드가 더 좋지 않을까요?

pkyung replied:
그렇게 하면 return도 안 쓸 수 있어서 좋은 것 같습니다. 수정하겠습니다.

ksy90101 commented on src/main/java/gift/model/Option.java:
```diff
@@ -53,4 +54,12 @@ public Long getQuantity() {
         return product;
     }
+
+    public void subtract(Long quantity) {
```
이부분에 대한 테스트 코드를 작성해보면 어떨까요?
모델의 테스트 코드도 한번 연습해보시면 좋을거 같아요.

pkyung replied:
네 적용해보겠습니다.

ksy90101 commented on src/test/java/gift/service/OptionServiceTest.java:
```diff
@@ -97,4 +96,18 @@ public void testMakeOption_ProductNotFound() {
         assertEquals("해당 id의 상품이 존재하지 않습니다.", thrown.getMessage());
     }
+
+    @Test
+    @DisplayName("subtract 성공 테스트")
+    void testSubtractSuccess() {
+        option.subtract(5L);
+        assertEquals(5L, option.getQuantity());
+    }
+
+    @Test
+    @DisplayName("quantity 정확히 0의 값으로 만드는 테스트")
+    void testSubtractExactQuantity() {
+        option.subtract(10L);
+        assertEquals(0L, option.getQuantity());
+    }
```
이건 옵션 테스트 아닌가요?

pkyung replied:
수정했습니다.

ksy90101 commented on src/main/java/gift/service/OptionService.java:
```diff
@@ -42,4 +43,11 @@ public Option makeOption(Long productId, OptionRequest request) {
     }
+
+    @Transactional
+    public void subtractQuantity(OptionSubtractRequest request) {
```
이부분에 대한 테스트코드 잘 되었는지 체크해보면 좋겠습니다.

ksy90101 reviewed (changes requested):
고생많으셨어요~
간단한 피드백을 남겼으니 확인해주세요!

pkyung committed:
Refactor: 멘토님 피드백 젹용

1. exception \n 적용
2. domain인 option 테스트의 subtract 테스트 적용했다.
3. servie의 subtractQuantity 테스트 적용했다.

ksy90101 commented on src/main/java/gift/model/Option.java:
```diff
@@ -53,4 +54,12 @@ public Long getQuantity() {
+        if (remainingQuantity < 0) {
+            throw new OutOfStockException("해당 옵션의 수량이 부족합니다.");
+        }
+        this.quantity -= quantity;
```
```suggestion
        this.quantity = remainingQuantity;
```

굳이 두번 계산할 필요가 없을거 같아서 변수 추출을 이야기 드린거였습니다.

ksy90101 reviewed (approved):
고생많으셨습니다~
5주차 미션에서 뵐게요!
홧팅입니다!
